src/elements/common/logger/Logger.js

```diff
@@ -0,0 +1,200 @@
+     * @param {string} name - the event name
+     * @returns {boolean} True if the event has already been fired
+     */
+    hasLoggedEvent(name: string): boolean {
+        return this.uniqueEvents.has(name);
+    }
+
+    /**
+     * Invokes the provided metric logging callback.
+     *
+     * @param {string} type - the type of the event
+     * @param {Object} data  - the event data
+     * @param {string} name - the name of the event
+     */
+    logMetric(type: MetricTypes, name: string, data: Object): void {
+        const metric: ElementsMetric = {
+            ...data,
+            component: this.props.source,
+            name,
+            timestamp: this.getTimestamp(),
+            sessionId: this.sessionId,
+            type,
+        };
+
+        this.props.onMetric(metric);
+    }
+
+    /**
+     * Logs a unique metric event. Prevents duplicate events from being logged in the session.
+     *
+     * @param {string} type - the type of the event
+     * @param {Object} data  - the event data
+     * @param {string} name - the name of the event
+     */
+    logUniqueMetric(type: MetricTypes, name: string, data: Object): void {
+        let eventName;
+        if (type !== METRIC_TYPE_ELEMENTS_LOAD_METRIC) {
+            if (!this.props.fileId) {
+                return;
+            }
+            eventName = this.createFileEventNameWith(name, this.props.fileId);
+        } else {
+            eventName = this.createEventName(name);
+        }
+
+        if (this.hasLoggedEvent(eventName)) {
+            return;
+        }
+
+        this.logMetric(type, name, data);
+        this.uniqueEvents.add(eventName);
+    }
+
+    /**
+     * Log a timing related metric.
+     *
+     * @param {string} type - the type of the event
+     * @param {Object} data  - the event data
+     * @param {string} name - the name of the event
+     * @param {boolean} isUnique  - true if the metric should be logged only once
+     * @returns {void}
+     */
+    logTimeMetric(type: MetricTypes, data: Object, name: string, isUnique: boolean): void {
+        if (!name) {
+            return;
+        }
+
+        if (isUnique) {
+            this.logUniqueMetric(type, name, data);
+        } else {
+            this.logMetric(type, name, data);
+        }
+    }
+
+    /**
+     * onMetric handler
+     *
+     * @param {string} type - the type of the event
+     * @param {Object} data  - the event data
+     * @param {string} name - the name of the event
+     * @param {boolean} isUnique  - true if the metric should be logged only once
+     * @returns {void}
+     */
+    handleMetric = (type: MetricTypes, data: Object, name?: string, isUnique?: boolean = true) => {
+        if (type === METRIC_TYPE_PREVIEW_METRIC) {
+            this.props.onMetric({
+                ...data,
+                type,
+            });
+        } else if (name) {
+            this.logTimeMetric(type, data, name, isUnique);
```

Comment:
Might want to standardize on either `type, data, name, ...` params ordering, or `type, name, data, ...` ordering, across this class